### PR TITLE
fix: Add percentencoding to special character `+`

### PR DESCRIPTION
**Pull request by harish1992**
As documented in CouchDB issues https://github.com/apache/couchdb/issues/2144 and https://github.com/apache/couchdb/issues/1830 creation of documents and attachments with unescaped `+` in `id` and attachment name is possible but requires them to be percentencoded while retrieving.

This PR provides a workaround to the issue #104 by percentencoding `+` and hence allowing creation,retrieval,updation and deletion of documents with `+` in their ids.

This change requires an update to Kitura-NIO 2.2 to pick changes from PR (https://github.com/IBM-Swift/Kitura-NIO/pull/199) to work, and hence won't work for anyone using Kitura-NIO with Swift 4.

## Description

Added `static func escape(url: String)` which takes `customAllowedCharacterSet` as argument to escape `+`
## Motivation and Context

This PR provides a workaround to this issue by percentencoding `+` and hence
allow creation,retrieval,updation and deletion of documents with `+` in their
ids.

Fixes the issue #104 .

## How Has This Been Tested?

Tested creation, retrieval, updation and deletion of CouchDb documents with `+` in their
`id`. Added Test case for creation, addition, retrieval and deletion of document attachments 
`+` in the attachment names.

## Checklist:

- [x] I have submitted a [CLA form](https://github.com/IBM-Swift/CLA)
- [x] If applicable, I have updated the documentation accordingly.
- [x] If applicable, I have added tests to cover my changes.

### CODE REVIEW

File: Tests/CouchDBTests/credentials.json [hunk 1]
```diff
@@ -1,4 +1,4 @@
 {
    "host": "127.0.0.1",
-   "port": 5984
+   "port": 5984,
```
Comment (djones6) [inline Tests/CouchDBTests/credentials.json]: Did you mean to do this?

File: Tests/CouchDBTests/DocumentCrudTests.swift [hunk 2]
```diff
@@ -44,10 +46,20 @@ class DocumentCrudTests: CouchDBTest {
                                     created_at: "Mon Aug 27 20:16:20 +0000 2012",
                                     favorited: false,
                                     value: "value2")
+    let myDocument4 = TypeADocument(_id: "654321+",
```
Comment (djones6) [inline Tests/CouchDBTests/DocumentCrudTests.swift]: Could you replace the literal ID string here with `documentId4` instead? (avoid duplicating strings) - same comment goes for all the documents defined here

File: Sources/CouchDB/CouchDBUtils.swift [hunk 3]
```diff
@@ -76,7 +76,35 @@ class CouchDBUtils {
...
+    // https://github.com/apache/couchdb/issues/2144 and https://github.com/apache/couchdb/issues/1830
+    // Workaround is to percent encode `+`. Following `customAllowedCharacterSet` variable escapes `+` in url.
+
+    private static let customAllowedCharacterSet =  NSCharacterSet(charactersIn:"\"#%/<>?@\\^`{|}+ ").inverted
+
+    /**
+     Transform the URL into escaped characters.
+
+     - note: URLs can only be sent over the Internet using the ASCII character set, so character escaping will
+     transform unsafe ASCII characters with a '%' followed by two hexadecimal digits.
```
Comment (djones6) [inline Sources/CouchDB/CouchDBUtils.swift]: This whole comment could simply be:
> Applies percent-encoding to a string, suitable for use as a CouchDB document path. In addition to characters normally encoded in a URL path, instances of the `+` character are also percent-encoded.

I don't think we need to explain why percent-encoding must be performed or its implementation - that is part of the HTTP specification.

**Comment by djones6** [comment]
Travis setup does not allow PR builds to work from a fork; here's a link to Harish's personal Travis build for https://github.com/harish1992/Kitura-CouchDB/commit/d2b404247b8cf412240c535ee3c5ad739909d00c:
https://travis-ci.org/harish1992/Kitura-CouchDB/builds/580062021

File: Package.swift [hunk 4]
```diff
@@ -23,7 +23,7 @@ import Foundation
 var kituraNetPackage: Package.Dependency
 
 if ProcessInfo.processInfo.environment["KITURA_NIO"] != nil {
-    kituraNetPackage = .package(url: "https://github.com/IBM-Swift/Kitura-NIO.git", from: "1.0.0")
+    kituraNetPackage = .package(url: "https://github.com/IBM-Swift/Kitura-NIO.git", from: "2.2.0")
```
Comment (djones6) [inline Package.swift:26]: I understand why we are making this change (to pick up https://github.com/IBM-Swift/Kitura-NIO/pull/199), but we should call it out in the PR description.

Note that due to the absence of the above, this fix won't work for anyone using Kitura-NIO with Swift 4 - although I expect that to affect little to no users in practice, as early -NIO adopters will likely be on Swift 5 already.

Comment (harish1992) [reply Package.swift:26]: @djones6  I have added this in PR

**Review by djones6 — APPROVED** [review]

**Pull request merged**